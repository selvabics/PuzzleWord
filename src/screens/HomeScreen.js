/**
 * This is Home screen
 * User can view their overall score
 * User can logout
 * List of category of puzzle will be shown
 * Progress of each category is displayed
 * User can select any category and start a new game
 * They have a option to see Leader board too
 */
import React, { useCallback, useState } from 'react';
import { StyleSheet, View, Text, TouchableOpacity, Alert, Image } from 'react-native';
import { useSelector } from 'react-redux';
import { responsiveWidth } from 'react-native-responsive-dimensions';

import Puzzles from '../utils/puzzles';
import CommonStyles from '../utils/styles';
import { navigate, popToTop } from '../router/navigation';
import { getCurrentUserSelector } from '../redux/slice/loginSlice';
import BackgroundWallper from '../components/backgroundWallpaper';
import ActionButton from '../components/actionButton';
import Category from '../components/category';
import { Logout, Gold } from '../assets';

const HomeScreen = () => {
    const [category, setCategory] = useState(undefined);
    const currentUser = useSelector(getCurrentUserSelector);

    const logout = () => {
        popToTop('loginScreen');
    };

    const goToLeaderBoardScreen = () => {
        navigate('leaderBoardScreen');
    }

    const goToPuzzleScreen = useCallback(() => {
        if (!category) {
            Alert.alert('Info', 'Please select category');
            return;
        }
        const level = currentUser.level[category.categoryId] || 0;
        if (level >= category.puzzles.length) {
            Alert.alert('Info', 'You have completed this category');
            return;
        }
        navigate('puzzleScreen', { category });
    }, [category])

    const selectCategory = (puzzle) => {
        setCategory(puzzle);
    };

    const renderCategory = (puzzle) => {
        const isSelectedCategory = puzzle.categoryId === category?.categoryId;
        const level = currentUser.level[puzzle.categoryId] || 0;
        const points = currentUser.points[puzzle.categoryId] || 0;
        return (
            <Category
                key={`${puzzle.categoryId}`}
                onPress={() => selectCategory(puzzle)}
                categoryName={puzzle.categoryName}
                isSelectedCategory={isSelectedCategory}
                level={level}
                rounds={puzzle.puzzles.length}
                points={points}
            />
        )
    }

    return (
        <BackgroundWallper>
            <View style={CommonStyles.containerView}>
                <View style={CommonStyles.headerView}>
                    <TouchableOpacity onPress={logout}>
                        <Image
                            style={styles.logoutIcon}
                            source={Logout}
                        />
                    </TouchableOpacity>
                    <Text style={CommonStyles.headerText}>Word Puzzle</Text>
                    <View style={styles.iconView}>
                        <Image
                            style={styles.goldIcon}
                            source={Gold}
                        />
                        <Text style={[CommonStyles.h4, { color: 'white' }]}>{currentUser.totalPoints}</Text>
                    </View>
                </View>
                <View style={styles.categoryContainer}>
                    {Puzzles.map(puzzle => (renderCategory(puzzle)))}
                    <ActionButton
                        title={'START'}
                        onPress={goToPuzzleScreen}
                    />
                </View>
                <View>
                    <ActionButton
                        title={'LEADER BOARD'}
                        onPress={goToLeaderBoardScreen}
                    />
                </View>
            </View>
        </BackgroundWallper>
    );
};

const styles = StyleSheet.create({

    categoryContainer: {
        alignItems: 'center',
        justifyContent: 'space-around',
    },
    iconView: {
        flexDirection: 'row',
        alignItems: 'center',
    },
    logoutIcon: {
        width: responsiveWidth(8),
        height: responsiveWidth(8),
    },
    goldIcon: {
        width: responsiveWidth(6),
        height: responsiveWidth(6),
        margin: 5,
    },
});

export default HomeScreen;
