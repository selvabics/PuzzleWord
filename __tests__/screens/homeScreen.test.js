import renderer from 'react-test-renderer';
import HomeScreen from '../../src/screens/HomeScreen';

import { Provider } from 'react-redux'
import configureStore from 'redux-mock-store';

jest.useFakeTimers();

describe('Testing login screen', () => {
    it('renders Login Screen', () => {
        const mockStore = configureStore()
        const currentUser = {
            email: '',
            level: {},
            points: {},
            totalPoints: 0,
        };
        const store = mockStore({ loginReducer: { currentUser } });
        const tree = renderer
            .create(<Provider store={store}><HomeScreen /></Provider>)
            .toJSON();
        expect(tree).toMatchSnapshot();
    });
});

