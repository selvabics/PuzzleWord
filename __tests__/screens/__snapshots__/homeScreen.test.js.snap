// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Testing login screen renders Login Screen 1`] = `
<View
  accessibilityIgnoresInvertColors={true}
  style={
    [
      {
        "flex": 1,
        "paddingBottom": 20,
        "paddingTop": 30,
      },
      {},
    ]
  }
>
  <Image
    source={
      {
        "testUri": "../../../src/assets/images/background.jpg",
      }
    }
    style={
      [
        {
          "bottom": 0,
          "left": 0,
          "position": "absolute",
          "right": 0,
          "top": 0,
        },
        {
          "height": undefined,
          "width": undefined,
        },
        undefined,
      ]
    }
  />
  <View
    style={
      {
        "alignItems": "center",
        "flex": 1,
        "justifyContent": "space-between",
      }
    }
  >
    <View
      style={
        {
          "alignItems": "center",
          "flexDirection": "row",
          "justifyContent": "space-between",
          "padding": 20,
          "width": "100%",
        }
      }
    >
      <View
        accessibilityState={
          {
            "busy": undefined,
            "checked": undefined,
            "disabled": undefined,
            "expanded": undefined,
            "selected": undefined,
          }
        }
        accessibilityValue={
          {
            "max": undefined,
            "min": undefined,
            "now": undefined,
            "text": undefined,
          }
        }
        accessible={true}
        collapsable={false}
        focusable={true}
        onClick={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          {
            "opacity": 1,
          }
        }
      >
        <Image
          source={
            {
              "testUri": "../../../src/assets/images/logout.jpg",
            }
          }
          style={
            {
              "height": 60,
              "width": 60,
            }
          }
        />
      </View>
      <Text
        style={
          {
            "color": "white",
            "fontSize": 45.89389937671454,
            "margin": 15,
            "textAlign": "center",
          }
        }
      >
        Word Puzzle
      </Text>
      <View
        style={
          {
            "alignItems": "center",
            "flexDirection": "row",
          }
        }
      >
        <Image
          source={
            {
              "testUri": "../../../src/assets/images/gold.jpg",
            }
          }
          style={
            {
              "height": 45,
              "margin": 5,
              "width": 45,
            }
          }
        />
        <Text
          style={
            [
              {
                "fontSize": 38.244916147262124,
              },
              {
                "color": "white",
              },
            ]
          }
        >
          0
        </Text>
      </View>
    </View>
    <View
      style={
        {
          "alignItems": "center",
          "justifyContent": "space-around",
        }
      }
    >
      <View
        accessibilityState={
          {
            "busy": undefined,
            "checked": undefined,
            "disabled": undefined,
            "expanded": undefined,
            "selected": undefined,
          }
        }
        accessibilityValue={
          {
            "max": undefined,
            "min": undefined,
            "now": undefined,
            "text": undefined,
          }
        }
        accessible={true}
        collapsable={false}
        focusable={true}
        onClick={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          {
            "opacity": 1,
          }
        }
      >
        <View
          collapsable={false}
          forwardedRef={[Function]}
          style={
            {
              "alignItems": "center",
              "backgroundColor": "rgba(255,255,255,0.4)",
              "borderRadius": 5,
              "borderWidth": 1,
              "flexDirection": "row",
              "height": 80.03999999999999,
              "justifyContent": "space-between",
              "margin": 22.5,
              "opacity": 0,
              "padding": 7.5,
              "transform": [
                {
                  "translateY": -800,
                },
              ],
              "width": 525,
            }
          }
        >
          <View
            style={
              [
                {
                  "alignItems": "center",
                  "justifyContent": "center",
                },
                {
                  "marginHorizontal": 7.5,
                },
              ]
            }
          >
            <Text
              style={
                {
                  "fontSize": 15.297966458904849,
                }
              }
            >
              Level
            </Text>
            <Text
              style={
                [
                  {
                    "fontSize": 22.94694968835727,
                  },
                  {
                    "fontWeight": "bold",
                  },
                ]
              }
            >
              0/10
            </Text>
          </View>
          <Text
            style={
              {
                "fontSize": 38.244916147262124,
              }
            }
          >
            Cities
          </Text>
          <View
            style={
              [
                {
                  "alignItems": "center",
                  "justifyContent": "center",
                },
                {
                  "marginHorizontal": 7.5,
                },
              ]
            }
          >
            <Text
              style={
                {
                  "fontSize": 15.297966458904849,
                }
              }
            >
              Score
            </Text>
            <Text
              style={
                [
                  {
                    "fontSize": 22.94694968835727,
                  },
                  {
                    "fontWeight": "bold",
                  },
                ]
              }
            >
              0
            </Text>
          </View>
        </View>
      </View>
      <View
        accessibilityState={
          {
            "busy": undefined,
            "checked": undefined,
            "disabled": undefined,
            "expanded": undefined,
            "selected": undefined,
          }
        }
        accessibilityValue={
          {
            "max": undefined,
            "min": undefined,
            "now": undefined,
            "text": undefined,
          }
        }
        accessible={true}
        collapsable={false}
        focusable={true}
        onClick={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          {
            "opacity": 1,
          }
        }
      >
        <View
          collapsable={false}
          forwardedRef={[Function]}
          style={
            {
              "alignItems": "center",
              "backgroundColor": "rgba(255,255,255,0.4)",
              "borderRadius": 5,
              "borderWidth": 1,
              "flexDirection": "row",
              "height": 80.03999999999999,
              "justifyContent": "space-between",
              "margin": 22.5,
              "opacity": 0,
              "padding": 7.5,
              "transform": [
                {
                  "translateY": -800,
                },
              ],
              "width": 525,
            }
          }
        >
          <View
            style={
              [
                {
                  "alignItems": "center",
                  "justifyContent": "center",
                },
                {
                  "marginHorizontal": 7.5,
                },
              ]
            }
          >
            <Text
              style={
                {
                  "fontSize": 15.297966458904849,
                }
              }
            >
              Level
            </Text>
            <Text
              style={
                [
                  {
                    "fontSize": 22.94694968835727,
                  },
                  {
                    "fontWeight": "bold",
                  },
                ]
              }
            >
              0/10
            </Text>
          </View>
          <Text
            style={
              {
                "fontSize": 38.244916147262124,
              }
            }
          >
            Foods
          </Text>
          <View
            style={
              [
                {
                  "alignItems": "center",
                  "justifyContent": "center",
                },
                {
                  "marginHorizontal": 7.5,
                },
              ]
            }
          >
            <Text
              style={
                {
                  "fontSize": 15.297966458904849,
                }
              }
            >
              Score
            </Text>
            <Text
              style={
                [
                  {
                    "fontSize": 22.94694968835727,
                  },
                  {
                    "fontWeight": "bold",
                  },
                ]
              }
            >
              0
            </Text>
          </View>
        </View>
      </View>
      <View
        accessibilityState={
          {
            "busy": undefined,
            "checked": undefined,
            "disabled": undefined,
            "expanded": undefined,
            "selected": undefined,
          }
        }
        accessibilityValue={
          {
            "max": undefined,
            "min": undefined,
            "now": undefined,
            "text": undefined,
          }
        }
        accessible={true}
        collapsable={false}
        focusable={true}
        onClick={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          {
            "opacity": 1,
          }
        }
      >
        <View
          collapsable={false}
          forwardedRef={[Function]}
          style={
            {
              "alignItems": "center",
              "backgroundColor": "rgba(255,255,255,0.4)",
              "borderRadius": 5,
              "borderWidth": 1,
              "flexDirection": "row",
              "height": 80.03999999999999,
              "justifyContent": "space-between",
              "margin": 22.5,
              "opacity": 0,
              "padding": 7.5,
              "transform": [
                {
                  "translateY": -800,
                },
              ],
              "width": 525,
            }
          }
        >
          <View
            style={
              [
                {
                  "alignItems": "center",
                  "justifyContent": "center",
                },
                {
                  "marginHorizontal": 7.5,
                },
              ]
            }
          >
            <Text
              style={
                {
                  "fontSize": 15.297966458904849,
                }
              }
            >
              Level
            </Text>
            <Text
              style={
                [
                  {
                    "fontSize": 22.94694968835727,
                  },
                  {
                    "fontWeight": "bold",
                  },
                ]
              }
            >
              0/10
            </Text>
          </View>
          <Text
            style={
              {
                "fontSize": 38.244916147262124,
              }
            }
          >
            Animals
          </Text>
          <View
            style={
              [
                {
                  "alignItems": "center",
                  "justifyContent": "center",
                },
                {
                  "marginHorizontal": 7.5,
                },
              ]
            }
          >
            <Text
              style={
                {
                  "fontSize": 15.297966458904849,
                }
              }
            >
              Score
            </Text>
            <Text
              style={
                [
                  {
                    "fontSize": 22.94694968835727,
                  },
                  {
                    "fontWeight": "bold",
                  },
                ]
              }
            >
              0
            </Text>
          </View>
        </View>
      </View>
      <View
        accessibilityState={
          {
            "busy": undefined,
            "checked": undefined,
            "disabled": undefined,
            "expanded": undefined,
            "selected": undefined,
          }
        }
        accessibilityValue={
          {
            "max": undefined,
            "min": undefined,
            "now": undefined,
            "text": undefined,
          }
        }
        accessible={true}
        collapsable={false}
        focusable={true}
        onClick={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          {
            "opacity": 1,
          }
        }
      >
        <View
          collapsable={false}
          forwardedRef={[Function]}
          style={
            {
              "alignItems": "center",
              "backgroundColor": "rgba(136,213,66,0.8)",
              "borderRadius": 5,
              "borderWidth": 1,
              "height": 80.03999999999999,
              "justifyContent": "center",
              "margin": 20,
              "opacity": 0,
              "padding": 7.5,
              "transform": [
                {
                  "scale": 0.3,
                },
              ],
              "width": 525,
            }
          }
        >
          <Text
            style={
              {
                "fontSize": 30.595932917809698,
                "fontWeight": "bold",
              }
            }
          >
            START
          </Text>
        </View>
      </View>
    </View>
    <View>
      <View
        accessibilityState={
          {
            "busy": undefined,
            "checked": undefined,
            "disabled": undefined,
            "expanded": undefined,
            "selected": undefined,
          }
        }
        accessibilityValue={
          {
            "max": undefined,
            "min": undefined,
            "now": undefined,
            "text": undefined,
          }
        }
        accessible={true}
        collapsable={false}
        focusable={true}
        onClick={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          {
            "opacity": 1,
          }
        }
      >
        <View
          collapsable={false}
          forwardedRef={[Function]}
          style={
            {
              "alignItems": "center",
              "backgroundColor": "rgba(136,213,66,0.8)",
              "borderRadius": 5,
              "borderWidth": 1,
              "height": 80.03999999999999,
              "justifyContent": "center",
              "margin": 20,
              "opacity": 0,
              "padding": 7.5,
              "transform": [
                {
                  "scale": 0.3,
                },
              ],
              "width": 525,
            }
          }
        >
          <Text
            style={
              {
                "fontSize": 30.595932917809698,
                "fontWeight": "bold",
              }
            }
          >
            LEADER BOARD
          </Text>
        </View>
      </View>
    </View>
  </View>
</View>
`;
